source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+import re
+from pathlib import Path
+from typing import \
+    List, \
+    Mapping, \
+    Tuple, \
+    Type
+from fab.artifact import \
+    Artifact, \
+    FileType, \
+    State, \
+    New, \
+    Unknown, \
+    Analysed, \
+    Compiled, \
+    Linked
+from fab.tasks import Task
+from fab.database import \
+    SqliteStateDatabase, \
+    FileInfoDatabase
+
+
+class PathMap(object):
+    def __init__(self,
+                 pattern: str,
+                 filetype: Type[FileType],
+                 state: Type[State]):
+        self._pattern = pattern
+        self._filetype = filetype
+        self._state = state
+
+    @property
+    def pattern(self):
+        return self._pattern
+
+    @property
+    def filetype(self):
+        return self._filetype
+
+    @property
+    def state(self):
+        return self._state
+
+    def match(self, path: Path) -> bool:
+        matched = False
+        if re.match(self.pattern, str(path)):
+            matched = True
+        return matched
+
+
+class Engine(object):
+    def __init__(self,
+                 workspace: Path,
+                 target: str,
+                 pathmaps: List[PathMap],
+                 taskmap: Mapping[
+                     Tuple[Type[FileType], Type[State]],
+                     Task]) -> None:
+        self._workspace = workspace
+        self._target = target
+        self._pathmaps = pathmaps
+        self._taskmap = taskmap
+        self._database = SqliteStateDatabase(workspace)
+
+    @property
+    def target(self) -> str:
+        return self._target
+
+    def process(self,
+                artifact: Artifact,
+                discovery,
+                objects,
+                lock) -> List[Artifact]:
+
+        new_artifacts = []
+        new_discovery = {}
+        new_objects = []
+        # Identify tasks that are completely new
+        if (artifact.state is New
+                and artifact.filetype is Unknown):
+            # Use the pathmap list to work out the
+            # filetype and starting state
+            new_artifact = None
+            for pathmap in self._pathmaps:
+                if pathmap.match(artifact.location):
+                    new_artifact = Artifact(artifact.location,
+                                            pathmap.filetype,
+                                            pathmap.state)
```

Comment:
There is probably a neater way of doing this by implementing `__special__` methods on the artifact. Maybe a way to use the `in` keyword. Something to bare in mind for subsequent changes.

Reply:
Ah, of course!  It was easy to change the method to `__contains__` and then we can happily have `if artifact.location in pathmap` which is much nicer - thanks :+1: